sources/api/netdog/src/main.rs

```diff
@@ -44,204 +51,133 @@ lazy_static! {
-    interface_family: InterfaceFamily,
-    data_file: PathBuf,
+    #[argh(subcommand)]
+    subcommand: SubCommand,
 }
 
-/// Stores fields extracted from a DHCP lease.
-#[derive(Debug, Deserialize)]
-struct LeaseInfo {
-    #[serde(rename = "ipaddr")]
-    ip_address: IpNet,
-    #[serde(rename = "dnsservers")]
-    dns_servers: BTreeSet<IpAddr>,
-    #[serde(rename = "dnsdomain")]
-    dns_domain: Option<String>,
-    #[serde(rename = "dnssearch")]
-    dns_search: Option<Vec<String>>,
+#[derive(FromArgs, PartialEq, Debug)]
+#[argh(subcommand)]
+enum SubCommand {
+    Install(InstallArgs),
+    Remove(RemoveArgs),
+    NodeIp(NodeIpArgs),
+    GenerateHostname(GenerateHostnameArgs),
+    SetHostname(SetHostnameArgs),
 }
 
-/// Informs the user about proper usage of the program and exits.
-fn usage() -> ! {
-    let program_name = env::args().next().unwrap_or_else(|| "program".to_string());
-    eprintln!(
-        r"Usage: {}
-            [ node-ip | install | remove ]
-
-            Required for 'install' and 'remove' subcommands:
-              -i INTERFACE_NAME
-              -t INTERFACE_TYPE
-              -f INTERFACE_FAMILY
-              DATA_FILE",
-        program_name
-    );
-    process::exit(2);
-}
+#[derive(FromArgs, PartialEq, Debug)]
+#[argh(subcommand, name = "install")]
+/// Write resolv.conf and current IP to disk
+struct InstallArgs {
+    #[argh(option, short = 'i')]
+    /// name of the network interface
+    interface_name: InterfaceName,
 
-/// Prints a more specific message before exiting through usage().
-fn usage_msg<S: AsRef<str>>(msg: S) -> ! {
-    eprintln!("{}\n", msg.as_ref());
-    usage();
+    #[argh(option, short = 't')]
+    /// network interface type
+    interface_type: InterfaceType,
+
+    #[argh(option, short = 'f')]
+    /// network interface family (ipv4/6)
+    interface_family: InterfaceFamily,
+
+    #[argh(positional)]
+    /// lease info data file
+    data_file: PathBuf,
+
+    #[argh(positional)]
+    // wicked might add `info` to the call to this program.  We don't do anything with it but must
+    // be able to parse the option to avoid failing
+    /// ignored
+    info: Option<String>,
 }
 
-/// Parses user arguments into a Subcommand and an Args structure.
-fn parse_args(args: env::Args) -> Result<(SubCommand, Option<Args>)> {
-    let mut iter = args.skip(1);
-    let value = iter
-        .next()
-        .unwrap_or_else(|| usage_msg("Did not specify command"));
-    let sub_command = serde_plain::from_str::<SubCommand>(&value)
-        .unwrap_or_else(|_| usage_msg(format!("Unknown command {}", value)));
-
-    // The `node-ip` subcommand doesn't require any arguments
-    if sub_command == SubCommand::NodeIp {
-        return Ok((sub_command, None));
-    };
+#[derive(FromArgs, PartialEq, Debug)]
+#[argh(subcommand, name = "remove")]
+// `wicked` may call `remove` with the below args and failing to parse them can cause an error in
+// `wicked`.
+/// Not implemented
```

Comment:
It is implemented at least as far as having a function that prints that it's not implemented - so I'd remove this comment.